app/javascript/lib/metrics-perf.js

```diff
@@ -138,3 +138,97 @@ export function setupWebVitalsLog() {
+function roundValues(props) {
+  Object.keys(props).forEach(key => {
+    props[key] = Math.round(props[key])
+  })
+  return props
+}
+
+/** Calculates generic performance timing metrics for API calls */
+export function calculatePerfTimes(perfTimes) {
+  const now = performance.now()
+
+  const plot = perfTimes.plot ? perfTimes.plot : 0
+
+  const perfEntry =
+    performance.getEntriesByType('resource')
+      .filter(entry => entry.name === perfTimes.url)[0]
+
+  const transfer = perfEntry.responseEnd - perfEntry.responseStart
+
+  const frontend = now - perfEntry.responseStart
+  const frontendOther = frontend - plot - perfTimes.parse - transfer
+
+  const backend = perfEntry.responseStart - perfEntry.requestStart
+
+  const full = now - perfEntry.startTime
+
+  const legacy = perfTimes.legacyBackend + plot
+
+  const compressedSize = perfEntry.encodedBodySize
+  const uncompressedSize = perfEntry.decodedBodySize
+  const compressionBytesDiff = uncompressedSize - compressedSize
+
+  const rawPerfProps = {
+    // Server + client timing
+    'perfTime:full': full, // Time from API call start to effect (e.g. plot) end
+
+    // Less precise, less complete times.  Retained for continuity.
+    // Old `perfTime` was measured from API request start to response end,
+    // which is very incomplete (lacks client times!) and less precise than
+    // using browsers' PerformanceEntry API.
+    'perfTime': legacy,
+    'perfTime:legacy': legacy,
+
+    // Server timing
+    'perfTime:backend': backend, // Time for server to process request
+
+    // Client timing
+    'perfTime:frontend': frontend, // Time from API call response start to effect (e.g. plot) end
+    'perfTime:frontend:transfer': transfer, // Time client took to download data from server
+    'perfTime:frontend:parse': perfTimes.parse, // Time to parse response body (currently only JSON)
+    'perfTime:frontend:other': frontendOther, // Total frontend time - accounted segments
+
+    // To answer questions about data sizes and compression
+    'perfTime:data:compressed-size': compressedSize, // Encoded response body size in bytes
+    'perfTime:data:uncompressed-size': uncompressedSize, // Decoded response body size in bytes
+    'perfTime:data:compression-bytes-diff': compressionBytesDiff // Absolute amount compressed
+  }
+
+  if (perfTimes.plot) {
+    rawPerfProps['perfTime:frontend:plot'] = perfTimes.plot
+  }
+  const perfProps = roundValues(Object.assign({}, rawPerfProps))
+
+  let compressionRatio = uncompressedSize / compressedSize
+  // Round to 2 digits, e.g. "3.44".  Number.EPSILON ensures numbers like 1.005 round correctly.
+  compressionRatio = Math.round((compressionRatio + Number.EPSILON) * 100) / 100
+  perfProps['perfTime:data:compression-ratio'] = compressionRatio // Relative amount compressed
+
+  // Accounts for `null`, '', 'non-empty string', etc.
+  const errorKeys = Object.keys(rawPerfProps).filter(k => isNaN(parseFloat(rawPerfProps[k])))
```

Comment:
As discussed, move this to above line 204, so it's closer to where the possible error occurred.

Reply:
Done in f1f273a.